src/digestif_conv.ml

```diff
@@ -0,0 +1,53 @@
+module Make (D : sig val digest_size : int end) = struct
+  let to_hex hash =
+    let res = Bytes.create (D.digest_size * 2) in
+
+    let chr x = match x with
+      | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 -> Char.chr (48 + x)
+      | _ -> Char.chr (97 + (x - 10)) in
+
+    for i = 0 to D.digest_size - 1
+    do
+      let v = Char.code (String.get hash i) in
+      Bytes.unsafe_set res (i * 2) (chr (v lsr 4));
+      Bytes.unsafe_set res (i * 2 + 1) (chr (v land 0x0F));
+    done;
+
+    Bytes.unsafe_to_string res
+
+  let fold f a s =
+    let r = ref a in
+    String.iter (fun x -> r := f !r x) s; !r
+
+  let of_hex hex =
+    let code x = match x with
+      | '0' .. '9' -> Char.code x - 48
+      | 'A' .. 'F' -> Char.code x - 55
+      | 'a' .. 'z' -> Char.code x - 87
+      | _ -> invalid_arg "of_hex" in
+
+    let wsp = function ' ' | '\t' | '\r' | '\n' -> true | _ -> false in
+
```

Comment:
Another attempt at minimizing usage of unsafe_ without sacrificing too much speed.
This should match your semantics (silent right-pad with zero-bytes, complain if you pass uneven amount of hex chars, ignore any whitespace in between characters):
```ocaml
let of_hex hex =
  let code x = match x with
      | '0' .. '9' -> Char.code x - 48
      | 'A' .. 'F' -> Char.code x - 55
      | 'a' .. 'z' -> Char.code x - 87
      | _ -> invalid_arg "of_hex" in
  let decode_char ch1 ch2 = Char.chr ((code ch1 lsl 4) lor (code ch2)) in
  let offset = ref 0 in
  let rec next have_first idx =
    if !offset + idx >= String.length v then
      '\x00' (* right-pad with 0x00 *)
    else match hex.[!offset + idx] with
    | ' ' | '\t' | '\r' | '\n' ->
      incr offset; next has_first idx
    | ch2 when has_first -> ch2
    | ch1 ->
      incr offset ;
      let ch2 = next true idx in
      if ch2 <> '\x00' then
        decode_char ch1 ch2
      else
        invalid_arg "of_hex: odd number of hex characters"
  in
  String.init D.digest_size (next false)
```

Reply:
Keep the code and we will make a new PR about that.